lib/puppet/type/aptly_purge.rb

```diff
@@ -26,10 +26,47 @@
     defaultto false
   end
 
+  newparam(:hold, :boolean => true, :parent => Puppet::Parameter::Boolean) do
+    defaultto false
+  end
+
   def generate
     outfile = @parameters[:debug] ? "/dev/stdout" : "/dev/null"
     package = Puppet::Type.type(:package)
 
+
+    # Using the RAL, divide the world into Catalog packages and not-Catalog
+    # packages.
+    managed_packages = []
+    unmanaged_packages = []
+
+    package.instances.select do |p|
+      p.provider.is_a?(Puppet::Type::Package::ProviderDpkg)
+    end.each do |r|
+      catalog_r = catalog.resource(r.ref)
+      if catalog_r.nil?
+        unmanaged_packages << r
+      else
+        managed_packages << catalog_r
+      end
+    end
+
+    managed_package_names = managed_packages.map(&:name)
+    unmanaged_package_names = unmanaged_packages.map(&:name)
+
+    if @parameters[:hold] then
+      # You can't hold a package that isn't installed yet, so this should
+      # really be done after all packages are installed.
+
+      versioned_package_names = managed_packages.select do |p|
+        # What we really want is to grab all packages with an explicit version
+        # This is a cheap reproduction of what we really want.
+        ![:latest, :absent, :present].include?(p.parameters[:ensure].value)
```

Comment:
Would it be better to check for a non-Symbol type? That might be a horrendous idea but it would protect against the eventuality that this list becomes non-exhaustive...